bin/create_cluster_config.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/python
+
+import sys
+import subprocess
+import json
+import argparse
+import yaml
+import pprint
+import os.path
+from ast import literal_eval
+
+# Init lists
+templates = []
+clusters = []
+instances = []
+bastions = []
+nodes = []
+masters = []
+output = []
+
+# Json output configuration
+
+
+class literal(str):
+    pass
+
+
+def literal_presenter(dumper, data):
+    return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
+
+# Function to fetch a value from the terraform dictionary
+
+
+def tf(key):
+    if terraform[key]['value']:
+        return terraform[key]['value']
+    else:
+        sys.exit(key + ' not found, exit.')
+
+
+# Dumper config
+yaml.SafeDumper.add_representer(literal, literal_presenter)
+yaml.SafeDumper.ignore_aliases = lambda *args: True
+
+# Parser config
+p = pprint.PrettyPrinter(indent=4)
+parser = argparse.ArgumentParser(description='Create Kubernetes clusters.')
+parser.add_argument(
+    'clusterconfig',
+    help='the base yaml file to create the cluster from')
+
+args = parser.parse_args()
+#print( 'loading file ' + args.clusterconfig)
+
+# Load the base kops yaml file located in the repo
+if os.path.isfile(args.clusterconfig):
+    stream = open(args.clusterconfig)
+    for template in yaml.load_all(stream):
+        templates.append(template)
+    stream.close()
+else:
+    sys.exit(args.clusterconfig + ' file not found, exit.')
+
+# Load the terraform outputs from the terraform command
+try:
+    terraform = json.loads(str(subprocess.check_output(
+        ['terraform', 'output', '-json']).decode('utf-8')))
+except subprocess.CalledProcessError as e:
+    print('error executing terraform command, exit.')
+
+# Populate variables from terraform output
+cluster_name = tf('cluster_domain_name')
+dns_zone = cluster_name
+kops_state_store = 's3://' + tf('kops_state_store') + '/' + cluster_name
+availability_zones = tf('availability_zones')
+master_public_name = 'api.' + cluster_name
+network_cidr = tf('network_cidr_block')
+topology = 'bastion.' + cluster_name
+internal_subnets = tf('internal_subnets')
+external_subnets = tf('external_subnets')
+
+# Organize in lists by kind and role
+for template in templates:
+    if template['kind'] == 'Cluster':
+        clusters.append(template)
+    if template['kind'] == 'InstanceGroup':
+        instances.append(template)
+        if template['spec']['role'] == 'Bastion':
+            bastions.append(template)
+        if template['spec']['role'] == 'Node':
+            nodes.append(template)
+        if template['spec']['role'] == 'Master':
+            masters.append(template)
+
+# Update all Cluster kind
+for template in clusters:
+    template['metadata'].update({'name': cluster_name})
```

Comment:
General python best practice note - it's not necessary to use the `update` method when setting a single key in a dict - `template['metadata']['name'] = cluster_name` does the same thing, and is shorter and easier to understand. `update()` is best used for setting several dict keys at once, so you should either use one `update()` call with several values, or directly set keys individually

Reply:
sounds good

Reply:
i changed this only for new keys
in cases where the key exists , using that method will override other values inside the key making it inviable
the .update method must be used in those cases